### PR TITLE
Loss of Precision in MST weight alteration

### PR DESCRIPTION
**Pull request by divyegala**

### CODE REVIEW

File: cpp/include/raft/sparse/mst/detail/mst_solver_inl.cuh [hunk 1]
```diff
@@ -116,7 +122,9 @@ MST_solver<vertex_t, edge_t, weight_t>::solve() {
   timer0 = duration_us(stop - start);
 #endif
 
-  Graph_COO<vertex_t, edge_t, weight_t> mst_result(2 * v - 2, stream);
+  auto n_expected_edges = symmetrize_output ? 2 * v - 2 : v - 1;
```
Comment (afender) [inline cpp/include/raft/sparse/mst/detail/mst_solver_inl.cuh:125]: This is not a valid expected number of edges for the spanning forest case.

File: cpp/include/raft/sparse/mst/detail/mst_solver_inl.cuh [hunk 2]
```diff
@@ -152,7 +160,12 @@ MST_solver<vertex_t, edge_t, weight_t>::solve() {
     timer3 += duration_us(stop - start);
 #endif
 
-    if (prev_mst_edge_count[0] == mst_edge_count[0]) {
+    auto curr_mst_edge_count = mst_edge_count[0];
+    RAFT_EXPECTS(curr_mst_edge_count <= n_expected_edges,
```
Comment (afender) [inline cpp/include/raft/sparse/mst/detail/mst_solver_inl.cuh:164]: This drops minimum spanning forest support

File: cpp/include/raft/sparse/hierarchy/detail/mst.cuh [hunk 3]
```diff
@@ -116,7 +116,7 @@ void connect_knn_graph(const raft::handle_t &handle, const value_t *X,
 
   // On the second call, we hand the MST the original colors
   // and the new set of edges and let it restart the optimization process
-  auto new_mst = raft::mst::mst<value_idx, value_idx, value_t>(
+  auto new_mst = raft::mst::mst<value_idx, value_idx, value_t, double>(
```
Comment (afender) [inline cpp/include/raft/sparse/hierarchy/detail/mst.cuh:119]: Using 64b is only helpful if `max_gap` is large compared to `max_weight`. Always using 64b may unnecessarily slow down the execution.

**Review by afender — CHANGES_REQUESTED** [review]
Dynamically selecting 64b only when needed would be a good improvement to consider

File: cpp/include/raft/sparse/mst/detail/mst_solver_inl.cuh (hunk 1 above)
Comment (afender) [reply cpp/include/raft/sparse/mst/detail/mst_solver_inl.cuh:125]: It should be seen as an upper bound

File: cpp/include/raft/sparse/hierarchy/detail/mst.cuh (hunk 3 above)
Comment (divyegala) [reply cpp/include/raft/sparse/hierarchy/detail/mst.cuh:119]: This is specifically to enable HDBSCAN (and remove errors in SLHC). While dynamic weight selection may help, MST is far from being the bottleneck for HDBSCAN, and we are trying to make HDBSCAN for this release so it's the path of least resistance for now

**Review by afender — APPROVED** [review]

**Comment by afender** [comment]
Thanks @divyegala !

**Comment by divyegala** [comment]
rerun tests

**Review by cjnolet — APPROVED** [review]
LGTM

**Comment by cjnolet** [comment]
@gpucibot merge

**Pull request merged**